src/Reduction.cpp

```diff
@@ -24,5 +29,36 @@ const std::vector<ReductionVariable> &ReductionDomain::domain() const {
     return contents.ptr->domain;
 }
 
+void ReductionDomain::where(const Expr &predicate) {
+    contents.ptr->predicates.push_back(predicate);
+}
+
+const std::vector<Expr> &ReductionDomain::predicates() const {
+    return contents.ptr->predicates;
+}
+
+std::vector<Expr> &ReductionDomain::predicates() {
+    return contents.ptr->predicates;
+}
+
+Expr ReductionDomain::and_predicates() const {
+	if (contents.ptr->predicates.empty()) {
+		return Expr();
+	}
+	Expr and_pred = contents.ptr->predicates[0];
+	for (size_t i = 1; i < contents.ptr->predicates.size(); ++i) {
+		and_pred = (and_pred && contents.ptr->predicates[i]);
+	}
+	return and_pred;
```

Comment:
I think this function could just be:
     return fold_right(contents.ptr->predicates, And::make)